### PR TITLE
8218685: jlink --list-plugins needs to be readable and tidy

### PR DESCRIPTION
**Pull request by igraves**
These changes update the jlink plugin command line documentation to tidy them up into a more canonical form.

The output generated by jlink from this change appears as follows:

```
> build/macosx-x64/images/jdk/bin/jlink --list-plugins

List of available plugins:
  --add-options <options>   Prepend the specified <options> string, which may
                            include whitespace, before any other options when
                            invoking the virtual machine in the resulting image.
  --compress <0|1|2>[:filter=<pattern-list>]
                            Compress all resources in the output image.
                            Level 0: No compression
                            Level 1: Constant string sharing
                            Level 2: ZIP.
                            An optional <pattern-list> filter can be
                            specified to list the pattern of 
                            files to be included.
  --dedup-legal-notices [error-if-not-same-content]
                            De-duplicate all legal notices.
                            If error-if-not-same-content is specified then
                            it will be an error if two files of the same
                            filename are different.
  --exclude-files <pattern-list>
                            Specify files to exclude.
                            e.g.: **.java,glob:/java.base/lib/client/**
  --exclude-jmod-section <section-name>
                            Specify a JMOD section to exclude.
                            Where <section-name> is "man" or "headers".
  --exclude-resources <pattern-list>
                            Specify resources to exclude.
                            e.g.: **.jcov,glob:**/META-INF/**
  --generate-jli-classes @filename
                            Specify a file listing the java.lang.invoke
                            classes to pre-generate. By default, this plugin
                            may use a builtin list of classes to pre-generate.
                            If this plugin runs on a different runtime version
                            than the image being created then code generation
                            will be disabled by default to guarantee 
                            correctness add ignore-version=true
                            to override this.
  --include-locales <langtag>[,<langtag>]*
                            BCP 47 language tags separated by a comma,
                            allowing
                            locale matching defined in RFC 4647.
                            e.g.: en,ja,*-IN
  --order-resources <pattern-list>
                            Order resources. 
                            e.g.: **/module-info.class,@classlist,
                            /java.base/java/lang/**
  --release-info <file>|add:<key1>=<value1>:<key2>=<value2>:...|del:<key list>
                            <file> option is to load release properties from
                            the supplied file.
                            add: is to add properties to the release file.
                            Any number of <key>=<value> pairs can be passed.
                            del: is to delete the list of keys in release file.
  --strip-debug             Strip debug information from the output image
  --strip-java-debug-attributes 
                            Strip Java debug attributes from
                            classes in the output image
  --strip-native-commands   Exclude native commands (such as java/java.exe)
                            from the image.
  --system-modules retainModuleTarget
                            Fast loading of module descriptors (always enabled)
  --vendor-bug-url <vendor-bug-url>
                            Override the vendor bug URL baked into the build.
                            The value of the system property
                            "java.vendor.url.bug" will be <vendor-url-bug>.
  --vendor-version <vendor-version>
                            Override the vendor version string baked into the
                            build,if any. The value of the system property
                            "java.vendor.version" will be <vendor-version>.
  --vendor-vm-bug-url <vendor-vm-bug-url>
                            Override the vendor VM bug URL baked 
                            into the build.  The URL displayed in VM error
                            logs will be <vendor-vm-bug-url>.
  --vm <client|server|minimal|all>
                            Select the HotSpot VM in the output image.
                            Default is all

For options requiring a <pattern-list>, the value will be a comma separated
list of elements each using one the following forms:
  <glob-pattern>
  glob:<glob-pattern>
  regex:<regex-pattern>
  @<filename> where filename is the name of a file containing patterns to be
              used, one pattern per line

```

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8218685](https://bugs.openjdk.java.net/browse/JDK-8218685): jlink --list-plugins needs to be readable and tidy

### Reviewers
 * [Mandy Chung](https://openjdk.java.net/census#mchung) (@mlchung - **Reviewer**) ⚠️ Review applies to fe79b36106740dfb1ec342e8b46cd9cf17a7feca
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to fe79b36106740dfb1ec342e8b46cd9cf17a7feca

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/305/head:pull/305`
`$ git checkout pull/305`

### CODE REVIEW

**Comment by mlchung** [comment]
Have you considered to sort them in alphabetical order of the plugin name?

**Comment by igraves** [comment]
Yes! I had intended to but it looks like I got hung up making sure non-documented plugins came last. Will push a change for this.

**Comment by AlanBateman** [comment]
Can we put a base abstract class in jdk.tools.jlink.internal.plugins that implements getUsage. That would avoid needing to have the same getUsage method in all the plugins.

The text in the value of some of the .usage keys looks like it will wrap significantly, can the values of release-info.usage and strip-native-debug-symbols.usage be re-formatted so that the usage is readable with a default terminal?

**Comment by igraves** [comment]
I'll shrink narrow those descriptions down to 80 characters.

As to the abstract class, I think it'd make sense to have that implement getUsage, getName, and getDescription -- essentially combine all of those very similar things in the same place.

**Comment by igraves** [comment]
Updated to consolidate documentation/resource-related methods into an abstract base class and shortened the mentioned usage bodies to 80 chars or less.

**Comment by AlanBateman** [comment]
AbstractPlugin might be a better name for the abstract class as it will likely define methods beyond access to name, description and usage. It also needs a copyright header.
StripNativeDebugSymbolsPlugin may have been missed in the update, it can extend the abstract class too.
AddResourcePlugin: the name field can be removed, also should be good to rename "n" to "name" so it's a bit clearer.
TaskHelper - the IDE may have messed up the imports when it was edited.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/TaskHelper.java [hunk 1]
```diff
@@ -30,18 +30,8 @@
...
 import java.text.MessageFormat;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.HashSet;
-import java.util.List;
-import java.util.Locale;
-import java.util.Map;
+import java.util.*;
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/TaskHelper.java:33]: I assume this change is unintentional.  Probably IDE converted the single-class imports to this.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/DocumentedPlugin.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package jdk.tools.jlink.internal.plugins;
+
+import jdk.tools.jlink.plugin.Plugin;
+
+public abstract class DocumentedPlugin implements Plugin {
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/DocumentedPlugin.java:5]: I agree with Alan that `AbstractPlugin` is better.

This should also provide the default implementation of `getArgumentsDescription()` such that the plugins don't need to implement.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/DocumentedPlugin.java [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+
+    private final String NAME;
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/DocumentedPlugin.java:7]: Nit: use lowercase for final instance field name.

File: src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/OrderResourcesPlugin.java [hunk 4]
```diff
@@ -46,23 +46,19 @@
  *
  * Order Resources plugin
  */
-public final class OrderResourcesPlugin implements Plugin {
+public final class OrderResourcesPlugin extends DocumentedPlugin {
     public static final String NAME = "order-resources";
```
Comment (mlchung) [inline src/jdk.jlink/share/classes/jdk/tools/jlink/internal/plugins/OrderResourcesPlugin.java:50]: This is not related to your change.   It can be private static final field.   In my past observation, this static final NAME field is defined in every plugin which isn't really a need for it (at least in the current implementation).   It might be good to consider the clean up to remove this static final field and instead call `getName()` to get the name.  Just a thought.

**Review by mlchung — COMMENTED** [review]
I suggest to update the PR description with the new `jlink --show-plugins` output.

**Review by mlchung — APPROVED** [review]
Looks good to me.  It may be useful to have a regression test to verify the plugin order of the output if possible.

**Comment by AlanBateman** [comment]
The new AbstractPlugin and each of the implementations extending it look good. The new usages messages look good too. A few minor formatting nits in a few places but otherwise I think good to go.
/approve

**Review by AlanBateman — APPROVED** [review]

**Comment by igraves** [comment]
Updated JLinkTest to assert alphabetical order. With that I will /integrate .
